xdsl/dialects/builtin.py

```diff
@@ -21,6 +21,76 @@
     from xdsl.printer import Printer
 
 
+@dataclass
+class ArrayOfConstraint(AttrConstraint):
+    """
+    A constraint that enforces an ArrayData whose elements all satisfy
+    the elem_constr.
+    """
+    elem_constr: AttrConstraint
+
+    def __init__(self, constr: Attribute | Type[Attribute] | AttrConstraint):
+        self.elem_constr = attr_constr_coercion(constr)
+
+    def verify(self, attr: Attribute) -> None:
+        if not isinstance(attr, Data):
+            raise Exception(f"expected data ArrayData but got {attr}")
+        for e in cast(ArrayAttr[Attribute], attr).data:
+            self.elem_constr.verify(e)
+
+
+_ArrayAttrT = TypeVar("_ArrayAttrT", bound=Attribute, covariant=True)
+
+
+@irdl_attr_definition
+class ArrayAttr(GenericData[List[_ArrayAttrT]]):
+    name = "array"
+
+    @staticmethod
+    def parse_parameter(parser: BaseParser) -> List[_ArrayAttrT]:
+        parser.parse_char("[")
+        data = parser.parse_list(parser.parse_optional_attribute)
```

Comment:
I am confused. does anyone else see the errors fron the reviewdog? Why are these not comments in the PR @superlopuh ?
Also, it seems that it is quite lost here, the parser certainly has this method...?

Reply:
Just for reference, this is code I moved up because I reference it, but it was there initially.

Reply:
Turns out the parser broke all of these functions. You'd think the parser has this method but it doesn't :) I'm not sure if there is one with a different name but same functionality. Would be good to fix soon

Reply:
Wait you are saying that we do not run this code in our tests? We do parse SymbolRefs, right?

Reply:
That is my current understanding

Reply:
Hmm okay... IMO, we should merge this then. The pyright issues have nothing to do with the actual changes, so there won't be more either...

Reply:
My understanding is that once the parser was changed, it broke these functions.
But these functions should have not existed in the first place, since they are never tested. They are never tested, because the parser has special handling for them anyway.
We should change that in a future PR.